File: Combinatorics/combinatorics.cpp
#include <iostream>
#include <vector>
#include <iterator>
using namespace std;

int factorial(int n){
    if(n == 0) return 1;
    else return n * factorial(n - 1);
}

int count_combinations(int n, int k){
    return factorial(n) / (factorial(k) * factorial(n - k));
}

int count_permutations(int n, int k){
    return factorial(n) / factorial(n - k);
}

vector <vector <int>> generate_combinations(vector <int> arr, int k){
    vector <vector <int>> result;
    for(int i = 0; i < count_combinations(arr.size(), k); i ++){
        vector <int> local_result(1, 0);
        int n = arr.size();
        int s = 0;
        for(int j = 1; j <= k; j++){
            int t = local_result[j - 1] + 1;
            while((t < (n - k + j)) && ((s + count_combinations(n - t, k - j) <= i))){
                s += count_combinations(n - t, k - j);
                t ++;
            }
            local_result.push_back(t);
        }
        for(int l = 1; l < local_result.size(); l ++){
            local_result[l] = arr[local_result[l] - 1];
        }
        result.push_back(vector <int>(local_result.begin() + 1, local_result.end()));
    }
    return result;
}

vector <vector <int>> generate_permutations(vector <int> arr, int k){
    vector <vector <int>> result;
    vector <vector <int>> m = generate_combinations(arr, k);
    for(auto a: m){
        vector <vector <int>> local_result;
        for(int i = 0; i < factorial(a.size()); i ++){
            int ind = i + 1;
            vector <int> local_local_result;
            vector <int> local_local_arr;
            for(auto r: a){
                local_local_arr.push_back(r);
            }
            int n = local_local_arr.size();
            for(int j = 1; j <= n; j ++){
                int f = factorial(n - j);
                int g = (ind + f - 1) / f;
                local_local_result.push_back(local_local_arr[g - 1]);
                auto iter = local_local_arr.cbegin();
                local_local_arr.erase(iter + g - 1);
                ind -= (g - 1) * f;
            }
            if(local_local_arr.size()){
                local_local_result.push_back(local_local_arr[0]);
            }
            local_result.push_back(local_local_result);
        }
        for(auto u: local_result){
            result.push_back(u);
        }
    }
    return result;
}

void print(vector <vector <int>> const &input)
{
    cout << "{ ";
    for(auto i: input){
        cout << "{ ";
        for(auto j: i){
            cout << j << ' ';
        }
        cout << "}";
    }
    cout << " }" << endl;
}

int main(){
    return 0;
}

File: Combinatorics/combinatorics.go
package main

func Factorial(n int) int {
	if n == 0 {
		return 1
	} else {
		return n * Factorial(n-1)
	}
}

func CountCombinations(n int, k int) int {
	return Factorial(n) / (Factorial(k) * Factorial(n-k))
}

func CountPermutations(n int, k int) int {
	return Factorial(n) / Factorial(n-k)
}

func GenerateCombinations(arr []int, k int) [][]int {
	result := [][]int{}
	for i := 0; i < CountCombinations(len(arr), k); i++ {
		localResult := []int{0}
		n := len(arr)
		s := 0
		for j := 1; j <= k; j++ {
			t := localResult[j-1] + 1
			for (t < (n - k + j)) && (s+CountCombinations(n-t, k-j) <= i) {
				s += CountCombinations(n-t, k-j)
				t++
			}
			localResult = append(localResult, t)
		}
		for l := 1; l < len(localResult); l++ {
			localResult[l] = arr[localResult[l]-1]
		}
		result = append(result, localResult[1:])
	}
	return result
}

func GeneratePermutations(arr []int, k int) [][]int {
	result := [][]int{}
	m := GenerateCombinations(arr, k)
	for _, a := range m {
		localResult := [][]int{}
		for i := 0; i < Factorial(len(a)); i++ {
			ind := i + 1
			localLocalResult := []int{}
			localLocalArr := []int{}
			for _, r := range a {
				localLocalArr = append(localLocalArr, r)
			}
			n := len(localLocalArr)
			for j := 1; j <= n; j++ {
				f := Factorial(n - j)
				g := (ind + f - 1) / f
				localLocalResult = append(localLocalResult, localLocalArr[g-1])
				localLocalArr = append(localLocalArr[:g-1], localLocalArr[g:]...)
				ind -= (g - 1) * f
			}
			if len(localLocalArr) > 0 {
				localLocalResult = append(localLocalResult, localLocalArr[0])
			}
			localResult = append(localResult, localLocalResult)
		}
		for _, u := range localResult {
			result = append(result, u)
		}
	}
	return result
}

func main() {

}
